### PR TITLE
Permit parallel CPU backend compilation (though currently disabled)

### PR DESCRIPTION
**Pull request by wsmoses**
Parallel compilation on CPU now available, though needs information from ISL to enable

### CODE REVIEW

**Comment by wsmoses** [comment]
This requires bumping halide and also the Tapir in the docker file as mentioned to @prigoyal

File: test/test_mapper_llvm.cc [hunk 1]
```diff
@@ -68,7 +68,7 @@ TEST(LLVMCodegen, MultiStmt) {
  {
    O1(i, j) +=! A(i, j, rk, rl) * B(i, j)
    O2(i, j) = C(i, j) * D(i, j)
-   O3(i, j) = O1(i, j) + O2(i, j)
+   O3(i, j) = O1(i, j) + O2(i, j)
```
Comment (ftynse) [inline test/test_mapper_llvm.cc]: Trailing whitespace

File: src/core/polyhedral/codegen_llvm.cc [hunk 2]
```diff
@@ -294,6 +303,88 @@ class CodeGen_TC : public Halide::Internal::CodeGen_X86 {
 
     value = sym_get(name);
   }
+
+ public:
+  void optimize_module() {
+    Halide::Internal::debug(3) << "Optimizing module\n";
```
Comment (ftynse) [inline src/core/polyhedral/codegen_llvm.cc]: Why use Halide's internal debug printer when we have our glog + flags?

File: src/core/polyhedral/codegen_llvm.cc [hunk 3]
```diff
@@ -451,6 +542,19 @@ class LLVMCodegen {
         llvm::BasicBlock::Create(llvmCtx, "loop_latch", function);
     auto* loopExitBB = llvm::BasicBlock::Create(llvmCtx, "loop_exit", function);
 
+    // TODO: integrate query ISL as to whether the relevant loop ought be
+    // parallelized
```
Comment (ftynse) [inline src/core/polyhedral/codegen_llvm.cc]: Open an issue?  :)

File: src/core/polyhedral/llvm_jit.cc [hunk 4]
```diff
@@ -35,11 +35,10 @@ Jit::Jit()
...
       DL_(TM_->createDataLayout()),
       objectLayer_([]() { return std::make_shared<SectionMemoryManager>(); }),
-      compileLayer_(objectLayer_, orc::SimpleCompiler(*TM_)),
-      optimizeLayer_(compileLayer_, [this](std::shared_ptr<Module> M) {
-        return optimizeModule(std::move(M));
-      }) {
+      compileLayer_(objectLayer_, orc::SimpleCompiler(*TM_)) {
   sys::DynamicLibrary::LoadLibraryPermanently(nullptr);
+  std::string err;
+  sys::DynamicLibrary::LoadLibraryPermanently("cilkrts", &err);
```
Comment (ftynse) [inline src/core/polyhedral/llvm_jit.cc]: Error message ignored?

File: src/core/polyhedral/llvm_jit.cc [hunk 5]
```diff
@@ -35,11 +35,10 @@ Jit::Jit()
     : TM_(EngineBuilder().selectTarget()),
       DL_(TM_->createDataLayout()),
       objectLayer_([]() { return std::make_shared<SectionMemoryManager>(); }),
-      compileLayer_(objectLayer_, orc::SimpleCompiler(*TM_)),
-      optimizeLayer_(compileLayer_, [this](std::shared_ptr<Module> M) {
-        return optimizeModule(std::move(M));
-      }) {
+      compileLayer_(objectLayer_, orc::SimpleCompiler(*TM_)) {
   sys::DynamicLibrary::LoadLibraryPermanently(nullptr);
```
Comment (ftynse) [inline src/core/polyhedral/llvm_jit.cc]: What's the point of keeping this?

File: src/core/polyhedral/codegen_llvm.cc (hunk 2 above)
Comment (wsmoses) [reply src/core/polyhedral/codegen_llvm.cc]: This function was taken verbatim from Halide [with the addition of one line]. Unfortunately, right now we need to do this but hopefully at one point Halide proper will support Tapir-like optimization [or permit me to just add the one line].

If you think it better, I can change to use our logging.

File: src/core/polyhedral/codegen_llvm.cc (hunk 3 above)
Comment (wsmoses) [reply src/core/polyhedral/codegen_llvm.cc]: Done, see #34

File: src/core/polyhedral/codegen_llvm.cc (hunk 2 above)
Comment (prigoyal) [reply src/core/polyhedral/codegen_llvm.cc]: I think it's better to use our own logging system. One of many reasons is also to control how we log things to terminal for python/C++ backend which right now needs to be solidified.

File: src/core/polyhedral/codegen_llvm.cc [hunk 6]
```diff
@@ -294,6 +303,88 @@ class CodeGen_TC : public Halide::Internal::CodeGen_X86 {
...
+    }
+
+    // We override PassManager::add so that we have an opportunity to
+    // blacklist problematic LLVM passes.
+    class MyFunctionPassManager : public llvm::legacy::FunctionPassManager {
+     public:
+      MyFunctionPassManager(llvm::Module* m)
+          : llvm::legacy::FunctionPassManager(m) {}
+      virtual void add(llvm::Pass* p) override {
+        Halide::Internal::debug(2)
```
Comment (prigoyal) [inline src/core/polyhedral/codegen_llvm.cc]: same comment as above :)

File: src/core/polyhedral/codegen_llvm.cc [hunk 7]
```diff
@@ -294,6 +303,88 @@ class CodeGen_TC : public Halide::Internal::CodeGen_X86 {
...
+#if LLVM_VERSION_MAJOR >= 5
+      module->print(llvm::dbgs(), nullptr, false, true);
+#else
+      module->dump();
+#endif
+    }
+
+    // We override PassManager::add so that we have an opportunity to
+    // blacklist problematic LLVM passes.
+    class MyFunctionPassManager : public llvm::legacy::FunctionPassManager {
```
Comment (prigoyal) [inline src/core/polyhedral/codegen_llvm.cc]: can you rename it to `LLVMPassManager`

Looking at `MyFunctionPassManager` seems like it is an unimplemented code :)

File: src/core/polyhedral/codegen_llvm.cc [hunk 8]
```diff
@@ -294,6 +303,88 @@ class CodeGen_TC : public Halide::Internal::CodeGen_X86 {
...
+
+    std::unique_ptr<llvm::TargetMachine> TM =
+        Halide::Internal::make_target_machine(*module);
+    module_pass_manager.add(llvm::createTargetTransformInfoWrapperPass(
+        TM ? TM->getTargetIRAnalysis() : llvm::TargetIRAnalysis()));
+    function_pass_manager.add(llvm::createTargetTransformInfoWrapperPass(
+        TM ? TM->getTargetIRAnalysis() : llvm::TargetIRAnalysis()));
+
+    llvm::PassManagerBuilder b;
+    b.OptLevel = 3;
```
Comment (prigoyal) [inline src/core/polyhedral/codegen_llvm.cc]: can you create constant variables for them in the header file and use that? we shouldn't hardcode magical numbers in the code :)  For example, in the header file:

```
static constexpr int kOptLevel = 3;
```

File: src/core/polyhedral/codegen_llvm.cc [hunk 9]
```diff
@@ -294,6 +303,88 @@ class CodeGen_TC : public Halide::Internal::CodeGen_X86 {
...
+    b.tapirTarget = new llvm::CilkABI();
+#if LLVM_VERSION_MAJOR >= 5
+    b.Inliner = llvm::createFunctionInliningPass(b.OptLevel, 0, false);
+#else
+    b.Inliner = llvm::createFunctionInliningPass(b.OptLevel, 0);
+#endif
+    b.LoopVectorize = true;
```
Comment (prigoyal) [inline src/core/polyhedral/codegen_llvm.cc]: why are they hardcoded like this? doesn't `PassManagerBuilder` have a proper constructor?

same comment for `b.SLPVectorize` below

File: src/core/polyhedral/codegen_llvm.cc [hunk 10]
```diff
@@ -294,6 +303,88 @@ class CodeGen_TC : public Halide::Internal::CodeGen_X86 {
...
+
+    // Run optimization passes
+    function_pass_manager.doInitialization();
+    for (llvm::Module::iterator i = module->begin(); i != module->end(); i++) {
+      function_pass_manager.run(*i);
+    }
+    function_pass_manager.doFinalization();
+    module_pass_manager.run(*module);
+
+    Halide::Internal::debug(3) << "After LLVM optimizations:\n";
```
Comment (prigoyal) [inline src/core/polyhedral/codegen_llvm.cc]: let's switch to our logging, the number 3 feels magical :)

File: src/core/polyhedral/codegen_llvm.cc [hunk 11]
```diff
@@ -294,6 +303,88 @@ class CodeGen_TC : public Halide::Internal::CodeGen_X86 {
...
+    class MyModulePassManager : public llvm::legacy::PassManager {
+     public:
+      virtual void add(llvm::Pass* p) override {
+        Halide::Internal::debug(2)
+            << "Adding module pass: " << p->getPassName().str() << "\n";
+        llvm::legacy::PassManager::add(p);
+      }
+    };
+
+    MyFunctionPassManager function_pass_manager(module.get());
```
Comment (prigoyal) [inline src/core/polyhedral/codegen_llvm.cc]: `functionPassManager`

File: src/core/polyhedral/codegen_llvm.cc [hunk 12]
```diff
@@ -294,6 +303,88 @@ class CodeGen_TC : public Halide::Internal::CodeGen_X86 {
...
+    MyModulePassManager module_pass_manager;
```
Comment (prigoyal) [inline src/core/polyhedral/codegen_llvm.cc]: `modulePassManager`

File: src/core/polyhedral/codegen_llvm.cc [hunk 13]
```diff
@@ -294,6 +303,88 @@ class CodeGen_TC : public Halide::Internal::CodeGen_X86 {
...
+
+    std::unique_ptr<llvm::TargetMachine> TM =
```
Comment (prigoyal) [inline src/core/polyhedral/codegen_llvm.cc]: change `TM` -> `targetMachine`

File: src/core/polyhedral/codegen_llvm.cc [hunk 14]
```diff
@@ -451,6 +542,19 @@ class LLVMCodegen {
         llvm::BasicBlock::Create(llvmCtx, "loop_latch", function);
     auto* loopExitBB = llvm::BasicBlock::Create(llvmCtx, "loop_exit", function);
 
+    // TODO: integrate query ISL as to whether the relevant loop ought be
+    // parallelized
+    bool parallel = false;
+
+    llvm::Value* SyncRegion = nullptr;
```
Comment (prigoyal) [inline src/core/polyhedral/codegen_llvm.cc]: how does this raw pointer gets freed?

**Comment by prigoyal** [comment]
@wsmoses I'll update the docker images for you today. The current tapir master should be fine?

File: src/core/polyhedral/codegen_llvm.cc (hunk 14 above)
Comment (wsmoses) [reply src/core/polyhedral/codegen_llvm.cc]: It's handled by LLVM

File: src/core/polyhedral/codegen_llvm.cc (hunk 9 above)
Comment (wsmoses) [reply src/core/polyhedral/codegen_llvm.cc]: Nope, constructor sets defaults, with no arguments

File: src/core/polyhedral/llvm_jit.cc [hunk 15]
```diff
@@ -29,17 +31,55 @@
...
 using namespace llvm;
 
+// Parse through ldconfig to find the path of a particular
+// shared library. This is an unfortunate way to have to
+// find it, but I couldn't immediately find something in
+// imported libraries that would resolve this for us.
+std::string find_library_path(std::string library) {
+  std::string command = "ldconfig -p | grep " + library;
+
+  FILE* fpipe = popen(command.c_str(), "r");
```
Comment (nicolasvasilache) [inline src/core/polyhedral/llvm_jit.cc]: @wsmoses  oh wow that is really painful.. can't we just set some macro at compile time like I've been doing for CUB?
https://github.com/facebookresearch/TensorComprehensions/blob/master/CMakeLists.txt#L113

**Pull request merged**